### PR TITLE
add support for config inheritance

### PR DESCRIPTION
**Pull request by aripringle**
Refs #183. This is a refactoring of #1013, with attempts to address the comments in that thread.

All credit goes to @1stvamp for the original code.

### CODE REVIEW

**Comment by alcohol** [comment]
It should probably be a root-only attribute (marked as such in documentation).

**Comment by far-blue** [comment]
+1

File: doc/04-schema.md [hunk 1]
```diff
@@ -824,4 +824,17 @@ Then "composer show -s" will give you `versions : * dev-latest-testing`.
 
 Optional.
 
+### extends <span>([root-only](04-schema.md#root-package))</span>
+
+Defines a parent JSON configuration filepath to a JSON file that this configuration extends.
+
+All properties from the parent JSON file are loaded and any property defined in the current
+file override (replace) those properties. This is done recursively so if the parent JSON file
+defines a parent that will be loaded (with any properties of it are also replaced) and so on.
```
Comment (cmfcmf) [inline doc/04-schema.md]: Do I understand correctly? It is *not* possible to only add some dependencies inside the child `composer.json` file, right? I.e. the parent requires `symfony/symfony` and the child requires `everything from parent` + `a/dependency`.

Comment (aripringle) [inline doc/04-schema.md]: > I.e. the parent requires symfony/symfony and the child requires everything from parent + a/dependency.

This is exactly how it works. I can try to clarify this in the documentation some.

Every first-level key in the composer.json file will have `array_replace` performed against it, effectively meaning that  keys from the child's `require` block will be appended to the parent's `require` block (keys of the same name will be overwritten).

While the child can overwrite values in the parent, you can't currently remove a parent value (you could technically try to set `symfony/symfony: null` or similar, but I don't think composer will treat that well).

Comment (cmfcmf) [inline doc/04-schema.md]: I was confused by 
> override (replace) those properties

It doesn't sound like extending to me.

File: src/Composer/Factory.php [hunk 2]
```diff
@@ -313,6 +313,31 @@ public function createComposer(IOInterface $io, $localConfig = null, $disablePlu
         return $composer;
     }
 
+    public static function readConfig($file, $mergeConfig = array())
```
Comment (alcohol) [inline src/Composer/Factory.php]: Why static by the way?

Comment (aripringle) [inline src/Composer/Factory.php]: It appeared that the calling method (createComposer) is sometimes called statically, and other Factory method calls are already being called statically from within that method.

File: doc/04-schema.md (hunk 1 above)
Comment (1stvamp) [inline doc/04-schema.md]: Alternatively you could use `inherits` to describe this functionality (`extends` was borrowed from `zc.buildout` terminology when I wrote the original PR).

**Comment by aripringle** [comment]
I wanted to explain my most recent commits, and open it up for discussion.

Anything defined as an object in composer.json will be "extended", meaning that new keys will be appended to the object, and existing keys will be overwritten (using `array_replace`). For example:

parent.json:
```json
{
  "require": {
    "php": ">=5.4.0",
    "monolog/monolog": "~1.0"
  }
}
```

child.json:
```json
{
  "require": {
    "monolog/monolog": "dev-master",
    "phpunit/phpunit": "dev-master"
  },
  "extends": "parent.json"
}
```

results in:
```json
{
  "require": {
    "php": ">=5.4.0",
    "monolog/monolog": "dev-master",
    "phpunit/phpunit": "dev-master"
  }
}
```

However, for numerically-indexed arrays, I opted to have the child values (if present) overwrite the parent values completely. I could opt to use `array_merge` here, always appending the child values to the parent, but this makes it impossible to override anything in the parent array (and could also allow for duplicate values in the array).

Any thoughts on the matter are welcome.

**Comment by ccoley** [comment]
This feature would be very useful. Any timeline on when it will be merged?

**Comment by czernitko** [comment]
An example of numerically indexed array is a list of repositories. We would like to use `extends` feature to add some customized bundles to Symfony project. Parent config file will be shared amongst all project installations, but for some installations we need to install additional customized bundles.
In such case, in child config file, we need to 
 - define new "require" item for custom bundle and 
 - add a new "repositories" record. 
If child values overwrote the parent values, it would mean that we have to copy parent values into child config file and add the new repositories. This basically means that we are not **extending**, but **replacing** the parent values. Therefore I would definitely vote for using `array_merge` to combine parent and child values in case of numerically indexed arrays.

I'm also very interested in an answer to ccoley's question :)

File: src/Composer/Factory.php (hunk 2 above)
Comment (Taluu) [inline src/Composer/Factory.php]: I don't see the point of creating a new method, as it is only called once...

File: src/Composer/Factory.php [hunk 3]
```diff
@@ -313,6 +313,36 @@ public function createComposer(IOInterface $io, $localConfig = null, $disablePlu
...
+                if (is_array($val)
+                    && array_key_exists($key, $config)
+                    && is_array($config[$key])
+                    && array_values($val) !== $val
+                    ) {
+                    $config[$key] = array_replace($config[$key], $val);
+                } else {
+                    $config[$key] = $val;
+                }
+            }
```
Comment (Taluu) [inline src/Composer/Factory.php]: isn't that basically the behaviour of `array_replace_recursive` ?

File: src/Composer/Factory.php [hunk 4]
```diff
@@ -313,6 +313,36 @@ public function createComposer(IOInterface $io, $localConfig = null, $disablePlu
         return $composer;
     }
 
+    public static function readConfig($file, $mergeConfig = array())
+    {
+        $config = $file->read();
+
+        if (!empty($mergeConfig)) {
```
Comment (Taluu) [inline src/Composer/Factory.php]: no need to test if `mergeConfig` is empty, the foreach will not iterate over it (as it is already empty)

File: src/Composer/Factory.php (hunk 2 above)
Comment (Taluu) [inline src/Composer/Factory.php]: nevermind, didn't see the recursive call

File: src/Composer/Factory.php [hunk 5]
```diff
@@ -313,6 +313,36 @@ public function createComposer(IOInterface $io, $localConfig = null, $disablePlu
...
+                    ) {
+                    $config[$key] = array_replace($config[$key], $val);
+                } else {
+                    $config[$key] = $val;
+                }
+            }
+        }
+
+        if (!empty($config['extends'])) {
+            $parent = new JsonFile($config['extends']);
```
Comment (Taluu) [inline src/Composer/Factory.php]: I dont think this is the correct path... If you input a `../parent.json`, wouldn't it load the file from the current directory (which is `/.../src/Composer`) ?

Comment (alcohol) [inline src/Composer/Factory.php]: No this is not really the current directory. Remember everything ends up wrapped in a Phar. It would be relative to the working directory of where you are running `composer`.

Comment (stof) [inline src/Composer/Factory.php]: @alcohol it will be resolved related to the current working directory (not the directory of this file), as any relative path in PHP. And this is what would be really confusing. A relative path in a composer.json should be resolved relatively to the composer.json file.

Comment (Taluu) [inline src/Composer/Factory.php]: And what if the path is not relative, or even on a distant remote filesystem ?

File: src/Composer/Factory.php (hunk 3 above)
Comment (aripringle) [inline src/Composer/Factory.php]: Almost, except the "else" here will also catch numerically-indexed arrays. If I do an array_replace in those, it would only replace matching keys, which could lead to strange-seeming behavior. For example:

```
"repositories": [
        { 
            // repository A...
        },
        { 
            // repository B...
        }
]
```

Extended with:

```
"repositories": [
        { 
            // repository C...
        }
]
```

If we use array_replace_recursive on this, we would end up with:

```
"repositories": [
        { 
            // repository C... (0th index was replaced)
        },
        { 
            // repository B...
        }
]
```

Instead, I opted to completely replace numerically-indexed arrays (this behavior is currently under discussion).

**Comment by aripringle** [comment]
@czernitko I was originally thinking that there might be a use case might that would require repositories to be replaced instead of extended. Without a more verbose syntax (such as allowing the user to specify which properties get extended), I thought this approach was a little more flexible.

However, after considering it a little more, extending the arrays as you're suggesting may be fine. For the repositories example specifically, it would probably make sense to `array_merge($childArray, $parentArray)`. This would allow the child repositories to take precedence over the parent repositories (in the case that they contain the same packages, for development/testing repositories for example).

**Comment by manfredjb** [comment]
Really useful, hope will be merged.

**Comment by composerfan8** [comment]
Hello. I am relatively new to composer, but could someone explain the rationale behind the bottom-up approach to inheritance using the 'extend' keyword, as opposed to a top-down approach ('include' in root composer.json) such as that found in composer-merge-plugin?

Looking at this from the perspective of a user with multiple subprojects as a part of a larger project, I think I must be missing something. It seems to me that requiring the 'extend' clause in these subprojects would make them less portable, where 'including' subproject composer.json would allow inheritance in the same way. This type of situation might occur in a monolithic repository situation.

Is the intent that subprojects should instead be referenced as composer packages in the main composer.json?

I may be missing the entire point here, but it would be nice to see a semi-detailed explanation of the thought process that went into this decision. Either way, it's great that progress is being made on the inheritance issue. Long live composer!

File: res/composer-schema.json [hunk 6]
```diff
@@ -434,6 +434,10 @@
             "items": {
                 "type": "string"
             }
+        },
+        "extends": {
+           "type": "string",
```
Comment (dashdanw) [inline res/composer-schema.json]: indentation?

**Comment by andrerom** [comment]
Cool!

Important step towards fixing distribution problems with composer based applications  *(maker and user both modifying same files)*. However any thoughts on seperating lock file and vendor dir as well? And should all relevant composer commands accept a -f <composer.json> like docker compose does?

**Comment by alcohol** [comment]
@andrerom please explain the purpose of a separate lock file and vendor directory? How do you see this working?

**Comment by andrerom** [comment]
Well, for me composer inheritance is a step towards solving a generic problem, "distribution problems". Example some symfony application, where users will be forced to migrate config, kernel, composer.json, .. and deal with conflicts as part of upgrades. This is somewhat pain, and you can easily make mistake if you're not that experienced and/or not fully awake. This is fixed by being able to have a dist version of composer.json.

As for the lock and vendor dir, this is more related to issues with distribution via docker images, however it is a separate issue and one I'm not fully able to put words on yet :) But if someone wants to discuss it, ping me on twitter or something.

**Comment by BernhardPosselt** [comment]
I may add to that that this is also important for anything that acts as a platform. For instance when writing an ownCloud app, you are defining your dependencies in a composer.json in your app folder. However ownCloud itself also ships multiple vendor libs via composer. I don't want to install guzzle in my app's vendor directory for instance when its already present in the main vendor directory.

**Comment by fabsalito** [comment]
In my opinion, this is a very useful feature. In my case, I work with multiple SSH identities and some clients has references in his composer.json to his own bundles in the way xxxxxx@bitbucket.org:client/yyyyy. It is complicated resolve this use case without the proposed feature.

**Comment by lvidal1** [comment]
Any updates on this?

**Comment by homerjam** [comment]
Yeesh, 4 years? LGTM?

Using https://github.com/wikimedia/composer-merge-plugin for now

**Comment by ruudboon** [comment]
Will this ever be part of composer?

**Comment by GrahamCampbell** [comment]
Is this PR dead?

**Comment by Pictor13** [comment]
Tok tok 🖖🏻, wondering the same . 
Was this stall for any reason?
Is its code still compatible with new Composer's versions?

Is it potentially merge-able or are there concerns from the core team about this feature?

**Comment by Seldaek** [comment]
Closing as per https://github.com/composer/composer/issues/183#issuecomment-716218186
